### PR TITLE
Load learning_resource objects into open search 

### PR DESCRIPTION
**Pull request by abeglova**
# What are the relevant tickets?
closes https://github.com/mitodl/mit-open/issues/38

# Description (What does it do?)
This PR loads data from learning_resources programs and courses into the open search index. 

There will be a follow up pr adding a search ui https://github.com/mitodl/mit-open/issues/75 that uses the search api

# How can this be tested?
On the main branch run `docker-compose run web ./manage.py backpopulate_xpro_data`, if you don't already have xpro courses and programs in your db

On this branch

run `docker-compose run web ./manage.py recreate_index --all`
Run 
`curl --location 'http://localhost:8063/api/v0/search/' \
--header 'Content-Type: application/json' \
--data '{  "post_filter": { "bool": { "must": [{"bool": { "should": [  { "term": {"resource_type": "program"}}]} }] }}}'`

and 
`curl --location 'http://localhost:8063/api/v0/search/' \
--header 'Content-Type: application/json' \
--data '{  "post_filter": { "bool": { "must": [{"bool": { "should": [  { "term": {"resource_type": "program"}}]} }] }}}'`

Verify that you see xpro course and program results

Run
`docker-compose run web ./manage.py backpopulate_xpro_data --delete`

Run the curl commands again and verify that no hits are returned

Run
`docker-compose run web ./manage.py backpopulate_xpro_data`

Run the curl commands again and verify that there are hits again

Also try running 

```
docker-compose run web ./manage.py update_index --all
```
And verify that there are no errors

### CODE REVIEW

File: learning_resources/etl/loaders.py [hunk 1]
```diff
@@ -357,7 +363,10 @@ def load_program(program_data, blocklist, duplicates, *, config=ProgramLoaderCon
             },
         )
 
-    # if not created and not program.published:
+    if not created and not program.learning_resource.published:
+        search_index_helpers.deindex_program(program.learning_resource)
+    elif program.learning_resource.published:
+        search_index_helpers.upsert_program(program.learning_resource.id)
```
Comment (mbertrand) [inline learning_resources/etl/loaders.py:369]: Wondering if we should try using `pluggy` mentioned by @rhysyngsun in [this RFC](https://github.com/mitodl/hq/discussions/1638) to kick off indexing tasks?  Or maybe that requires more discussion and this is a viable fallback for now until that is finalized.

File: learning_resources_search/serializers.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    instructors = serializers.SerializerMethodField()
+
+    def get_instructors(self, instance):
+        """
+        Get a list of instructor names for the course run
+        """
+        return [
+            (instructor.full_name or f"{instructor.first_name} {instructor.last_name}")
+            for instructor in instance.instructors.all()
+        ]
```
Comment (mbertrand) [inline learning_resources_search/serializers.py:30]: Can we apply this format to `LearningResourceRunSerializer` instead, so they share the same format in search results?

File: learning_resources_search/serializers.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+class OSLearningResourceSerializer(LearningResourceSerializer):
+    """Serializer for LearningResources for Search"""
+
+    topics = serializers.SerializerMethodField()
+
+    runs = OSLearningResourceRunSerializer(read_only=True, many=True, allow_null=True)
+
+    def get_topics(self, instance):
+        """Return the list of topics"""
+        return [topic.name for topic in instance.topics.all()]
```
Comment (mbertrand) [inline learning_resources_search/serializers.py:42]: Ditto - is it possible to have `topics` in the same format across both search and learning_resources?

File: learning_resources/etl/loaders_test.py [hunk 4]
```diff
@@ -57,24 +57,17 @@ def mock_duplicates(mocker):
...
+            "learning_resources_search.search_index_helpers.upsert_course"
         ),
-        delete_podcast=mocker.patch("search.search_index_helpers.deindex_podcast"),
-        delete_podcast_episode=mocker.patch(
-            "search.search_index_helpers.deindex_podcast_episode"
+        delete_course=mocker.patch(
+            "learning_resources_search.search_index_helpers.deindex_course"
         ),
-        index_run_content_files=mocker.patch(
-            "search.search_index_helpers.index_run_content_files"
```
Comment (mbertrand) [inline learning_resources/etl/loaders_test.py:77]: Future PR for contentfiles?

File: learning_resources_search/constants.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            "slug": {"type": "keyword"},
+            "availability": {"type": "keyword"},
+            "semester": {"type": "keyword"},
+            "year": {"type": "keyword"},
+            "start_date": {"type": "date"},
+            "end_date": {"type": "date"},
+            "enrollment_start": {"type": "date"},
+            "enrollment_end": {"type": "date"},
+            "instructors": {"type": "text"},
+            "prices": {"type": "scaled_float", "scaling_factor": 100},
```
[the remark this answers is not in the document]
Comment (abeglova) [reply learning_resources_search/constants.py:92]: Huh weird - I do not get this error when I run recreate index locally

[the remark this answers is not in the document]
Comment (abeglova) [reply learning_resources_search/constants.py:92]: And course-v1:xPRO+LASERx4 is in my search index

File: learning_resources_search/serializers.py (hunk 2 above)
Comment (mbertrand) [reply learning_resources_search/serializers.py:30]: PS, other than `topics` and `instructors` as mentioned above, the API and search responses for an individual program seem identical, which is great 👍

File: learning_resources_search/constants.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbertrand) [reply learning_resources_search/constants.py:92]: Got it to work after deleting all `LearningResource` objects, running `recreate_index`, then `backpopulate_xpro_data`, then `recreate_index` again.  As pointed out in slack, was likely due to the presence of obsolete `course_catalog` indices in my local instance.

File: learning_resources/etl/loaders_test.py (hunk 4 above)
Comment (abeglova) [reply learning_resources/etl/loaders_test.py:77]: yes!

File: learning_resources/etl/loaders.py (hunk 1 above)
Comment (abeglova) [reply learning_resources/etl/loaders.py:369]: I think we should try out pluggy but I don't want it to block this pr - maybe we can merge this for now and have a follow up pr that switches over to use pluggy?

Comment (mbertrand) [reply learning_resources/etl/loaders.py:369]: sounds good to me

File: learning_resources_search/search_index_helpers.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    try_with_retry_as_task(tasks.upsert_course, course_id)
+
+
+def deindex_course(learning_resource_obj):
+    """
+    Run a task to delete an ES Course document
+
+    Args:
+        learning_resource_obj (course_catalog.models.Course): A LearningReasoruce object
```
Comment (mbertrand) [inline learning_resources_search/search_index_helpers.py:45]:
```python
 learning_resource_obj (learning_resources.models.LearningResource): A LearningResource object of resource_type Course
 ```

File: learning_resources_search/search_index_helpers.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    try_with_retry_as_task(tasks.upsert_program, learning_resource_id)
+
+
+def deindex_program(learning_resource_obj):
+    """
+    Run a task to delete an ES Program document
+
+    Args:
+        program_obj (course_catalog.models.Program): A Program object
```
Comment (mbertrand) [inline learning_resources_search/search_index_helpers.py:72]: ```program_obj (learning_resource.models.LearningResource): A LearningResource object with resource_type Program```

**Review by mbertrand — CHANGES_REQUESTED** [review]
Just a couple minor docstring comments, also do you think adjusting the topics/instructors serialization to be the same between search and api is doable?

File: learning_resources_search/urls.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+"""URLs for search"""
+from django.urls import re_path
+
+from learning_resources_search.views import SearchView
+
+urlpatterns = [
+    re_path(r"api/v0/search/", SearchView.as_view(), name="search"),
```
Comment (mbertrand) [inline learning_resources_search/urls.py:7]: Change to `api/v1/search/` to be consistent with the learning_resources api

**Review by mbertrand — APPROVED** [review]
LGTM, just update the API version to match the new learning_resources api version (1 instead of 0)

**Pull request merged**